src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,359 @@
+        } else {
+            command_JumpBack.setEnabled(false);
+        }
+        
+        if (jumpedPosStack.length > 0) {
+            command_JumpFwd.setEnabled(true);
+        } else {
+            command_JumpFwd.setEnabled(false);
+        }
+    }
+    
+    /**
+    * Function to check existence of a file entry
+    * @private
+    */
+    function _checkIfExist(entry) {
+        var deferred = new $.Deferred(),
+            fileEntry = FileSystem.getFileForPath(entry.file);
+
+        if (entry.inMem) {
+            var indxInWS = MainViewManager.findInWorkingSet(entry.paneId, entry.file);
+            // Remove entry if InMemoryFile is not found in Working set
+            if (indxInWS === -1) {
+                deferred.reject();
+            } else {
+                deferred.resolve();
+            }
+        } else {
+            fileEntry.exists(function (err, exists) {
+                if (!err && exists) {
+                    deferred.resolve();
+                } else {
+                    deferred.reject();
+                }
+		    });
+	    }
+
+        return deferred.promise();
+    }
+        
+   /**
+    * Prototype to capture a navigation frame and it's various data/functional attributues
+    */
+    function NavigationFrame(editor, selectionObj) {
+        this.cm = editor._codeMirror;
+        this.file = editor.document.file._path;
+        this.inMem = editor.document.file.constructor.name === "InMemoryFile" ? true : false;
+        this.paneId = editor._paneId;
+        this.uId = (new Date()).getTime();
+        this.selections = [];
+        this.bookMarkIds = [];
+        this._createMarkers(selectionObj.ranges);
+        this._bindEditor(editor);
+    }
+    
+   /**
+    * Binds the lifecycle event listner of the editor for which this frame is captured
+    */
+    NavigationFrame.prototype._bindEditor = function (editor) {
+        var self = this;
+        editor.on("beforeDestroy", function () {
+            self._backupSelectionRanges();
+            self.cm = null;
+            self.bookMarkIds = null;
+        });
+    };
+    
+   /**
+    * Function to create CM TextMarkers for the navigated positions/selections.
+    * This logic is required to ensure that the captured navaigation positions 
+    * stay valid and contextual even when the actual document text mutates.
+    * The mutations which are handled here :
+    * -> Addition/Deletion of lines before the captured position
+    * -> Addition/Updation of characters in the captured selection
+    */
+    NavigationFrame.prototype._createMarkers = function (ranges) {
+        var range, index, bookMark;
+        this.bookMarkIds = [];
+        for (index in ranges) {
+            range = ranges[index];
+            // 'markText' has to used for a non-zero length position, if current selection is 
+            // of zero length use bookmark instead.
+            if (range.anchor.line === range.head.line && range.anchor.ch === range.head.ch) {
```

Comment:
If there is a cursor or selection, it can never be null and if there are no selections/cursor then the ranges would be an empty array.